lib/rgb_matrix/engine.ex

```diff
@@ -1,54 +1,52 @@
+require Logger
+
 defmodule RGBMatrix.Engine do
   @moduledoc """
-  Renders [`Animation`](`RGBMatrix.Animation`)s and outputs
-  [`Frame`](`RGBMatrix.Frame`)s to be displayed.
+  Renders [`Animation`](`RGBMatrix.Animation`)s and outputs colors to be
+  displayed by [`Paintable`](`RGBMatrix.Paintable`)s.
   """
 
   use GenServer
 
+  alias Layout.LED
   alias RGBMatrix.Animation
 
   defmodule State do
     @moduledoc false
-    defstruct [:animation, :paintables]
+    defstruct [:leds, :animation, :paintables, :last_frame, :timer]
   end
 
   # Client
 
   @doc """
   Start the engine.
 
-  This module registers its process globally and is expected to be started by
-  a supervisor.
+  This module registers its process globally and is expected to be started by a
+  supervisor.
 
   This function accepts the following arguments as a tuple:
-  - `initial_animation` - The animation that plays when the engine starts.
-  - `paintables` - A list of modules to output `RGBMatrix.Frame` to that implement
-      the `RGBMatrix.Paintable` behavior. If you want to register your paintables
-      dynamically, set this to an empty list `[]`.
+  - `leds` - The list of LEDs to be painted on.
+  - `initial_animation` - The Animation type to initialize and play when the
+    engine starts.
+  - `paintables` - A list of modules to output colors to that implement the
+    `RGBMatrix.Paintable` behavior. If you want to register your paintables
+    dynamically, set this to an empty list `[]`.
   """
-  @spec start_link({initial_animation :: Animation.t(), paintables :: list(module)}) ::
+  @spec start_link(
+          {leds :: [LED.t()], initial_animation_type :: Animation.type(),
+           paintables :: list(module)}
+        ) ::
           GenServer.on_start()
-  def start_link({initial_animation, paintables}) do
-    GenServer.start_link(__MODULE__, {initial_animation, paintables}, name: __MODULE__)
+  def start_link({leds, initial_animation_type, paintables}) do
+    GenServer.start_link(__MODULE__, {leds, initial_animation_type, paintables}, name: __MODULE__)
   end
 
   @doc """
-  Play the given animation.
-
-  Note that the animation can be played synchronously by passing `:false` for the `:async` option. However, only
-  looping (animations with `:loop` >= 1) animations may be played this way. This is to ensure that the caller is not
-  blocked forever.
+  Sets the given animation as the currently active animation.
   """
-  @spec play_animation(animation :: Animation.t(), opts :: keyword()) :: :ok
-  def play_animation(animation, opts \\ []) do
-    async? = Keyword.get(opts, :async, true)
-
-    if async? do
-      GenServer.cast(__MODULE__, {:play_animation, animation})
-    else
-      GenServer.call(__MODULE__, {:play_animation, animation})
-    end
+  @spec set_animation(animation_type :: Animation.type(), opts :: keyword()) :: :ok
+  def set_animation(animation_type) do
```

Comment:
:warning: (recommended)

I recommend we keep this as `play_animation`, as well as keep its arg an `%Animation{}` rather than an animation type.

Regarding the function name, when we iterate on the engine `play` will sound familiar to other transport controls we may want, like `stop_animation` and `pause_animation`. I also realized the return value of `play_animation` and `stop_animation` could be the animation that was playing, giving the application a way to hang on to this state if it wants the animation to pick up where it left off rather than resetting. This also prevents outside components from trying to reach into the engine to get this info.

Regarding the animation type arg, like I said previously, I think it would be better if the engine didn't need to be responsible for initializing animations. Just give it initialized animations to act on.

Reply:
I can keep the name if you really want, but to me it feels more like you're "setting the engine's current animation" rather than telling the animation to play something.  Some animations don't necessarily "play" (either right away or at all)...

As for the change from being a struct vs the type, I had to make this change for various reasons that I couldn't get around at the time.  If you can find a way later to undo this, I'm fine with it, but for now let's keep it this way.

Reply:
We can dive into this later. This is fine for now. :+1: